File: 0206-reverse-linked-list/0206-reverse-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

// Recursive method
class Solution {
public:
    ListNode* reverse(ListNode *curr, ListNode *prev){
        if(!curr) return prev;
        ListNode *temp = curr->next;
        curr->next = prev;
        prev = curr;
        curr = temp;
        prev = reverse(curr,prev);
        return prev;
    }
    ListNode* reverseList(ListNode* head) {
        if(!head || !head->next) return head;
        ListNode *prev = NULL;
        return reverse(head,prev);
    }
};


// Iterative method 

// class Solution {
// public:
//     ListNode* reverseList(ListNode* head) {
//         if(!head || !head->next) return head;
//         ListNode *curr = head, *prev = NULL;
//         while(curr){
//             ListNode *temp = curr->next;
//             curr->next = prev;
//             prev = curr;
//             curr = temp;
//         }
//         return prev;
        
//     }
// };

File: 0206-reverse-linked-list/0206-reverse-linked-list.cs
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int val=0, ListNode next=null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */
public class Solution {

    public ListNode Reverse(ListNode curr, ListNode prev){
        if(curr == null) return prev;
        ListNode temp = curr.next;
        curr.next = prev;
        prev = curr;
        curr = temp;
        return Reverse(curr, prev);
    }

    public ListNode ReverseList(ListNode head) {
        if(head == null || head.next == null) return head;

        ListNode prev = null;
        return Reverse(head, prev);
    }
}